src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/Connection/LoginHandler.cs

```diff
@@ -0,0 +1,636 @@
+
+            await TdsLogin(context, isAsync, ct).ConfigureAwait(false);
+
+            // If the workflow being used is Active Directory Authentication and server's prelogin response
+            // for FEDAUTHREQUIRED option indicates Federated Authentication is required, we have to insert FedAuth Feature Extension
+            // in Login7, indicating the intent to use Active Directory Authentication for SQL Server.
+            static bool ShouldRequestFedAuth(LoginHandlerContext context)
+            {
+                bool IsEntraIdAuthInConnectionString = context.ConnectionOptions.Authentication == SqlAuthenticationMethod.ActiveDirectoryPassword
+                                                || context.ConnectionOptions.Authentication == SqlAuthenticationMethod.ActiveDirectoryInteractive
+                                                || context.ConnectionOptions.Authentication == SqlAuthenticationMethod.ActiveDirectoryDeviceCodeFlow
+                                                || context.ConnectionOptions.Authentication == SqlAuthenticationMethod.ActiveDirectoryServicePrincipal
+                                                || context.ConnectionOptions.Authentication == SqlAuthenticationMethod.ActiveDirectoryManagedIdentity
+                                                || context.ConnectionOptions.Authentication == SqlAuthenticationMethod.ActiveDirectoryMSI
+                                                || context.ConnectionOptions.Authentication == SqlAuthenticationMethod.ActiveDirectoryDefault
+                                                || context.ConnectionOptions.Authentication == SqlAuthenticationMethod.ActiveDirectoryWorkloadIdentity
+                                                // Since AD Integrated may be acting like Windows integrated, additionally check _fedAuthRequired
+                                                || (context.ConnectionOptions.Authentication == SqlAuthenticationMethod.ActiveDirectoryIntegrated);
+
+                return IsEntraIdAuthInConnectionString && context.FedAuthNegotiatedInPrelogin
+                                || context.AccessTokenCallback != null;
+            }
+        }
+
+        private async ValueTask TdsLogin(LoginHandlerContext context, bool isAsync, CancellationToken ct)
+        {
+            // TODO: Password Change
+
+            context.TdsStream.PacketHeaderType = TdsStreamPacketType.Login7;
+
+            // Fixed length of the login record
+            int length = TdsEnums.SQL2005_LOG_REC_FIXED_LEN;
+
+            Debug.Assert(TdsEnums.MAXLEN_CLIENTINTERFACE >= context.ClientInterfaceName.Length, "cchCltIntName can specify at most 128 unicode characters. See Tds spec");
+
+            // Calculate the fixed length
+            checked
+            {
+                length += context.CalculateLoginRecordLength() * 2;
+                if (context.UseFeatureExt)
+                {
+                    length += 4;
```

Comment:
Can you push this logic block also in CalculateLoginRecordLength() ?

Reply:
done